enhancements/network/egress-qos.md

```diff
@@ -0,0 +1,187 @@
+  - "@trozet"
+  - "@danwinship"
+  - "@tssurya"
+approvers:
+  - TBD
+creation-date: 2022-02-16
+last-updated: 2022-02-27
+status: implementable
+---
+
+# OVN Pods Egress DSCP QoS
+
+## Summary
+
+Not all traffic has the same priority, and when there is contention for bandwidth, there should be a mechanism for objects outside the cluster to prioritize the traffic.
+To enable this, we will use Differentiated Services Code Point (DSCP) which allows us to classify packets by setting a 6-bit field in the IP header, effectively marking the priority of a given packet relative to other packets as "Critical", "High Priority", "Best Effort" and so on.
+
+By introducing a new CRD `EgressQoS`, users could specify a DSCP value for packets originating from pods on a given namespace heading to a specified CIDR.
+The CRD will be Namespaced, with one resource allowed per namespace.
+The resources will be watched by ovn-k, which in turn will configure OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#QoS_TABLE). 
+
+## Motivation
+
+Telco customers require support for DSCP marking capability for some of their 5G applications, giving some pods precedence over others.
+The QoS markings will be consumed and acted upon by objects outside of the OpenShift cluster to optimize traffic flow throughout their networks.
+
+### Goals
+
+- Provide a mechanism for users to set DSCP on egress traffic coming from specific namespaces.
+
+### Non-Goals
+
+- Ingress QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth QoS mechanism.
+
+- The DSCP marking does not need to be handled or acted upon by OpenShift, just added to selected headers.
+
+- Marking East/West traffic, exposing the DSCP value from the inner packet to the outer geneve packet.
+
+## Proposal
+
+To achieve egress DSCP marking on pods, we introduce a new namespace-scoped CRD `EgressQoS` which allows specifying a set of QoS rules, each has a DSCP value and a destination CIDR. Traffic coming from pods on the namespace heading to each destination CIDR will be marked with the corresponding DSCP value.
+
+### Implementation Details/Notes/Constraints
+
+A new API `EgressQoS` under the `k8s.ovn.org/v1` version will be added to `pkg/crd`.
+
+A new controller in OVN-K will watch `EgressQoS` and `Pod` objects, which will create the relevant QoS objects in OVN and result in the necessary flows to be programmed in OVS. By listing the pods in the namespace these QoS rules will be attached only to the relevant node local switches.
+
+For example, assuming there's a single pod `app1` in namespace `default` on node `node1` and the following `EgressQoS` is created:
+
+```yaml
+kind: EgressQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 46
+    dstCIDR: 0.0.0.0/0
+  - dscp: 30
+    dstCIDR: 1.2.3.4/32
```

Comment:
why do we only match on destination CIDR and then assume the source is all addresses in this namespace? I think it would be better to add srcCIDR option as well so that only certain pods may be selected.

Reply:
Wouldn't srcCIDR be a bit complicated for the user? Adding a PodSelector per rule is possible with the current design but I left it out for now as it seems that it is not a requirement.

Reply:
Actually revisiting my conversation with the customer, pod selector is very much a requirement. We want to be able to say `Pod with X label gets DSCP marking of Y for all of its traffic` as a functional MVP. namespace selector can come later perhaps and if we are making the CRD namespace scoped, then we will be OK.

Reply:
Added a PodSelector field for the QoS rules, and a bit about how we can use AddressSets to make sure matching pods are marked with the right DSCP.